diff --git a/veritas/driver/driver.go b/veritas/driver/driver.go
index a91d6ff..b929d65 100644
--- a/veritas/driver/driver.go
+++ b/veritas/driver/driver.go
@@ -2,11 +2,12 @@ package driver
 
 import (
 	"context"
+	"fmt"
 
 	"google.golang.org/grpc"
 
 	pbv "hybrid/proto/veritas"
-	"hybrid/tso"
+	tso "hybrid/tso_zookeeper"
 	"hybrid/veritas/db"
 )
 
@@ -18,6 +19,8 @@ type Driver struct {
 }
 
 func Open(serverAddr, tsoAddr, signature string) (*Driver, error) {
+	fmt.Println("Using Zookeeper TSO ...")
+
 	cc, err := grpc.Dial(serverAddr, grpc.WithInsecure())
 	if err != nil {
 		return nil, err
